File: new_backend/express-backend/src/routes/ingestion.js
const express = require('express');
const router = express.Router();
const Joi = require('joi');
const axios = require('axios');

// Validation schemas
const startIngestionSchema = Joi.object({
  agentId: Joi.string().required(),
  urls: Joi.array().items(Joi.string().uri()).default([]),
  s3Urls: Joi.array().items(Joi.string()).default([])
});

// Middleware to validate tenant access
const validateTenantAccess = (req, res, next) => {
  const tenantId = req.headers['x-tenant-id'] || req.query.tenantId;
  if (!tenantId) {
    return res.status(400).json({ error: 'Tenant ID required' });
  }
  req.tenantId = tenantId;
  next();
};

// Start document ingestion
router.post('/start', validateTenantAccess, async (req, res) => {
  try {
    const { error, value } = startIngestionSchema.validate(req.body);
    if (error) {
      return res.status(400).json({ error: error.details[0].message });
    }

    const prisma = req.app.get('prisma');
    const { agentId, urls, s3Urls } = value;

    // Verify agent belongs to tenant
    const agent = await prisma.agent.findFirst({
      where: {
        id: agentId,
        user: { id: req.tenantId }
      }
    });

    if (!agent) {
      return res.status(403).json({ error: 'Access denied' });
    }

    // Create documents in database
    const documents = [];
    for (const url of urls) {
      const doc = await prisma.document.create({
        data: {
          url: url,
          agentId: agentId,
          status: 'pending'
        }
      });
      documents.push(doc);
    }

    for (const s3Url of s3Urls) {
      const doc = await prisma.document.create({
        data: {
          s3Path: s3Url,
          agentId: agentId,
          status: 'pending'
        }
      });
      documents.push(doc);
    }

    // Trigger ingestion service
    const ingestionServiceUrl = process.env.FASTAPI_URL || 'http://localhost:8001';
    const ingestionResponse = await axios.post(`${ingestionServiceUrl}/ingest`, {
      tenantId: req.tenantId,
      agentId: agentId,
      urls: urls,
      s3_urls: s3Urls
    });

    res.json({
      jobId: ingestionResponse.data.job_id,
      documents: documents,
      status: 'processing'
    });
  } catch (error) {
    console.error('Error starting ingestion:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Check ingestion status
router.get('/status/:jobId', validateTenantAccess, async (req, res) => {
  try {
    const { jobId } = req.params;

    const ingestionServiceUrl = process.env.FASTAPI_URL || 'http://localhost:8001';
    const response = await axios.get(`${ingestionServiceUrl}/status/${jobId}`);

    res.json(response.data);
  } catch (error) {
    console.error('Error checking ingestion status:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get ingestion jobs for agent
router.get('/jobs', validateTenantAccess, async (req, res) => {
  try {
    const { agentId } = req.query;

    if (!agentId) {
      return res.status(400).json({ error: 'Agent ID required' });
    }

    const prisma = req.app.get('prisma');

    // Verify agent belongs to tenant
    const agent = await prisma.agent.findFirst({
      where: {
        id: agentId,
        user: { id: req.tenantId }
      }
    });

    if (!agent) {
      return res.status(403).json({ error: 'Access denied' });
    }

    const redis = req.app.get('redis');
    const pattern = `job:${req.tenantId}:${agentId}:*`;
    const keys = await redis.keys(pattern);

    const jobs = [];
    for (const key of keys) {
      const jobId = key.split(':')[3];
      const status = await redis.get(`job:${jobId}`);
      const progress = await redis.get(`job:${jobId}:progress`);

      jobs.push({
        jobId: jobId,
        status: status,
        progress: progress ? parseInt(progress) : 0
      });
    }

    res.json(jobs);
  } catch (error) {
    console.error('Error fetching ingestion jobs:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: python/backend/app/routes/ingestion.py
"""
/api/ingestion routes — mirrors Express src/routes/ingestion.ts
"""
from typing import Optional
from fastapi import APIRouter, Depends, Query, Request
from fastapi.responses import JSONResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
import httpx

from app.database import get_db
from app.auth import AuthContext, get_auth
from app.models import Agent, Document
from app.config import settings

router = APIRouter()


@router.post("/start")
async def start_ingestion(body: dict, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    agent_id = body.get("agentId", "knowledge_base")
    urls = body.get("urls", [])
    s3_urls = body.get("s3Urls", [])

    result = await db.execute(select(Agent).where(Agent.id == agent_id, Agent.tenantId == auth.tenant_id))
    agent = result.scalar_one_or_none()
    if not agent:
        result2 = await db.execute(select(Agent).where(Agent.tenantId == auth.tenant_id).limit(1))
        agent = result2.scalar_one_or_none()
    if not agent:
        return JSONResponse({"error": "No agents found for this tenant"}, status_code=403)

    documents = []
    for url in urls:
        doc = Document(url=url, agentId=agent.id, tenantId=auth.tenant_id, status="pending")
        db.add(doc)
        await db.flush()
        documents.append({"id": doc.id, "url": doc.url, "status": doc.status})

    for s3_url in s3_urls:
        doc = Document(s3Path=s3_url, agentId=agent.id, tenantId=auth.tenant_id, status="pending")
        db.add(doc)
        await db.flush()
        documents.append({"id": doc.id, "s3Path": doc.s3Path, "status": doc.status})

    await db.commit()

    job_id = None
    try:
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.post(
                f"{settings.FASTAPI_URL}/ingest",
                json={"tenantId": auth.tenant_id, "agentId": agent.id, "urls": urls, "s3_urls": s3_urls},
            )
            job_id = resp.json().get("job_id")
    except Exception:
        pass

    return {"jobId": job_id, "documents": documents, "status": "processing"}


@router.get("/status/{job_id}")
async def ingestion_status(job_id: str):
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.get(f"{settings.FASTAPI_URL}/status/{job_id}")
            return resp.json()
    except Exception:
        return JSONResponse({"error": "Failed to fetch status"}, status_code=500)


@router.get("/jobs")
async def ingestion_jobs(agentId: str = Query(...), auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    r = await db.execute(select(Agent).where(Agent.id == agentId, Agent.tenantId == auth.tenant_id))
    if not r.scalar_one_or_none():
        return JSONResponse({"error": "Access denied"}, status_code=403)
    # Simplified: return empty list since we don't have Redis job tracking
    return []
